psc-publish/Main.hs

```diff
@@ -196,18 +197,35 @@ getResolvedDependencies :: [PackageName] -> PrepareM [(PackageName, Version)]
 getResolvedDependencies declaredDeps = do
   depsBS <- fromString <$> readProcess' "bower" ["list", "--json", "--offline"] ""
 
-  deps <- catchLeft (parse asBowerResolvedDependencies depsBS)
-                    (internalError . JSONError FromBowerList)
+  -- Check for undeclared dependencies
+  toplevels <- catchJSON (parse asToplevelDependencies depsBS)
+  warnUndeclared declaredDeps toplevels
 
-  warnUndeclared declaredDeps (map fst deps)
+  deps <- catchJSON (parse asResolvedDependencies depsBS)
   handleDeps deps
 
   where
-  asBowerResolvedDependencies ::
-    Parse BowerError [(PackageName, DependencyStatus)]
-  asBowerResolvedDependencies =
-    key "dependencies"
-      (eachInObjectWithKey (parsePackageName . T.unpack) asDependencyStatus)
+  catchJSON = flip catchLeft (internalError . JSONError FromBowerList)
+
+-- | Extracts all dependencies and their versions from
+--   `bower list --json --offline`
+asResolvedDependencies :: Parse BowerError [(PackageName, DependencyStatus)]
+asResolvedDependencies = nubBy ((==) `on` fst) <$> go
+  where
+  go =
+    fmap (fromMaybe []) $
+      keyMay "dependencies" $
+        (++) <$> eachInObjectWithKey (parsePackageName . T.unpack)
+                                     asDependencyStatus
+             <*> (concatMap snd <$> eachInObject asResolvedDependencies)
```

Comment:
I just checked this - if you create a bower package with cyclic dependencies and then run `bower list --json --offline`,  it prints a dependency tree where packages are not listed where they would appear as children of themselves. So if A depends on B depends on A, the dependency tree will just look like `A -> B`. So I am pretty sure this is fine.